### DISCUSSION TITLE
Random Selecting Images by Groups

### DISCUSSION THREAD

**Discussion by wangzhuoranivy** [jspsych/jsPsych#876] [Q&A] [answered]
Hi. I need to code an experiment with its stimuli being a large number of images divided in 7 groups. In each block, four random images from each group should be presented, meaning 28 images from 7 different groups, the order is also random. I have been thinking for a while and tried several approaches but couldn't get around with it using the alternate-groups because it uses all the images in the group. Could anyone help with that? Thanks a lot in advance. 

Also I'm also wondering if the timeline variables can be named differently by the types like
timeline_variables: [
        { typea: 'person-1.jpg'},
        { typea: 'person-2.jpg'},
        { typeb: 'person-3.jpg'},
        { typeb: 'person-4.jpg',}
    ],

Or should I create an array of images for each group as the timeline variables?

**Comment by jenniekim00** [comment, marked as answer]
My project deals with randomly picking from a large group of images (150 total) which are separated into 3 groups (50 total in each). My approach that currently works includes:

1.  create different array of images for each group
2. create new arrays that is defined by using the jspsych.randomwith(orwithout)replacement code to pull from each of the arrays
3. create another array that conjoins arrays that randomly pulled images from each group, for example you can use the .concat method
4. insert the stimuli as a function with the last array in mind, so that it does this for every time you want to use it for a trial

Also heads up, if your images are different sizes, you might have to resize them externally and/or adjust the code, which leads to more technical expertise (i got help for resizing my different images in the code, so I can't really explain it but it was definitely a struggle aha).

Hope this helps!

**Reply by wangzhuoranivy** [reply]
Hi Jennie. Thanks a lot for your help. I'm wondering once I have an array of all images I want to display, how to make it a function as the stimuli. It would be great if you could give me some hints on that.

**Reply by jenniekim00** [reply]
Hi! I realized I meant to say create a function to call on in the 'on finish' and not for 'stimuli. There might be a better method than my approach, but here's an example of how I structured my code:

```javascript

var group1_images = [
  'img/1.png',
  'img/2.png',
  'img/3.png',
  'img/4.png',
  'img/5.png',
  'img/6.png',
  ]
//repeat code above for your 7 groups
//rename each array according to your group categories
//insert your specific image src 

var arraySize = 4 
//setting this as a separate variable, in case you want to change it and not have to edit every line of code following this

//creating new arrays of images, pulling 4 random images from each group
var array1 = jsPsych.randomization.sampleWithoutReplacement(group1_images, arraySize)
var array2 = jsPsych.randomization.sampleWithoutReplacement(group2_images, arraySize)
var array3 = jsPsych.randomization.sampleWithoutReplacement(group3_images, arraySize)
var array4 = jsPsych.randomization.sampleWithoutReplacement(group4_images, arraySize)
var array5 = jsPsych.randomization.sampleWithoutReplacement(group5_images, arraySize)
var array6 = jsPsych.randomization.sampleWithoutReplacement(group6_images, arraySize)
var array7 = jsPsych.randomization.sampleWithoutReplacement(group7_images, arraySize)

//combining all the 28 images into one array
var stimuli_images = array1.concat(array2, array3, array4, array5, array6, array7)

//Shuffling the array if you don't want all the same category images to be lined up next to each other
var ArrayOfImages = jsPsych.randomization.shuffle(stimuli_images)

//function to use to make sure your ArrayOfImages is being refreshed for each trial
var updateArrayOfImages = function(){

  arraySize = 4 
  array1 = jsPsych.randomization.sampleWithoutReplacement(group1_images, arraySize)
  array2 = jsPsych.randomization.sampleWithoutReplacement(group2_images, arraySize)
  array3 = jsPsych.randomization.sampleWithoutReplacement(group3_images, arraySize)
  array4 = jsPsych.randomization.sampleWithoutReplacement(group4_images, arraySize)
  array5 = jsPsych.randomization.sampleWithoutReplacement(group5_images, arraySize)
  array6 = jsPsych.randomization.sampleWithoutReplacement(group6_images, arraySize)
  array7 = jsPsych.randomization.sampleWithoutReplacement(group7_images, arraySize)

  stimuli_images = array1.concat(array2, array3, array4, array5, array6, array7)

  ArrayOfImages = jsPsych.randomization.shuffle(stimuli_images)
}

//creating trial to call on the images
 var recall = {
 
    type: 'vsl-grid-scene', //this plugin is helpful for displaying grid of images
    stimuli: function(){
    	return ArrayOfImages
    },
    
    on_finish: function(){
    	updateArrayOfImages()
    }

```

**Reply by jenniekim00** [reply]
you can also customize the vsl-grid-scene plugin and insert choice buttons, you can probably copy and paste that portion of code from the image-button-response plugin

**Comment by wangzhuoranivy** [comment]
Just tried this with reference to #762 but had an error of
Uncaught TypeError: Cannot read property 'trial' of undefined


var pic_scr = {
	type: 'image-button-response',
	stimulus: jsPsych.timelineVariable('trial'),
	data: {screen: 'image'},
	trial_duration: 5000,
        choices: ['Angry', 'Disgusted', 'Fear', 'Happy', 'Neutral', 'Sad', 'Surprised'],
	response_ends_trial: true,
	on_finish: function (data) {
		data.trialnum = trialnum;
                data.blocknum = blocknum;
	}
};
var trial_stimuli = [
       //first group
	{ trial: 'STIMULI/KDEF_F/ANGER_F/F01ANS_A_M.jpg' },
	{ trial: 'STIMULI/KDEF_F/ANGER_F/F02ANS_A_M.png' },
	{ trial: 'STIMULI/KDEF_F/ANGER_F/F03ANS_B_M.png' },
        //second group
	{ trial: 'STIMULI/KDEF_F/ANGER_F/F20ANS_B.jpg' },
	{ trial: 'STIMULI/KDEF_F/ANGER_F/F21ANS_B.jpg' },
	{ trial: 'STIMULI/KDEF_F/ANGER_F/F22ANS_A.jpg' },
        //third group
	{ trial: 'STIMULI/KDEF_F/DISGUST_F/F01DIS_B_M.png' },
        { trial: 'STIMULI/KDEF_F/DISGUST_F/F02DIS_A_M.png' },
        { trial: 'STIMULI/KDEF_F/DISGUST_F/F03DIS_B_M.png' }
];

var block = {
            timeline: [ready_scr, pic_scr, blank_scr],
            timeline_variables: trial_stimuli,
            // specify the randomization conditions
            sample: {
                type: 'custom',
                fn: function () {
                    //3 groups
		    var groups = [[0, 1, 2], [3, 4, 5], [6, 7, 8]];
                    var order = jsPsych.randomization.shuffle(groups); // shuffle groups
                    var selected_items = [];
                    for (var i = 0; i < 3; i++) { // iterate through 3 groups
				var new_items = jsPsych.randomization.sampleWithReplacement(order[i], 2); //select two items from each 
                                group
				selected_items.concat(new_items);
		    }
		    var final_items = jsPsych.randomization.shuffle(selected_items);
                    return final_items;
                  }
};

**Reply by conbainbridge** [reply]
A couple edits for var block:

When setting up your var new_items, using selected_items.concat isn't working since the array is initialized empty - you can change this to selected_items.push. Then, you will need to return an array that does not have arrays nested inside it. You could set this up like so:

```
var shuffled_items = jsPsych.randomization.shuffle(selected_items)
var final_items = [
          shuffled_items[0][0],
          shuffled_items[0][1],
          shuffled_items[1][0],
          shuffled_items[1][1],
          shuffled_items[2][0],
          shuffled_items[2][1]
];
```

Hope this helps!

(edits for clearer formatting of code)

**Comment by wangzhuoranivy** [comment]
Thank you both I just figured it out! :)
